8305626: Kitchensink7D.java failing because of native memory exhausting with ZGC

coleenp opened the pull request:
Use a ConcurrentHashTable to associate Method* with jmethodID, instead of an indirection.  JNI is deprecated in favor of using Panama to call methods, so I don't think we're concerned about performance going forward.  JVMTI uses a lot of jmethodIDs but there aren't any performance tests for JVMTI.

Tested with tier1-4, 5-7 in progress.

dcubed-ojdk commented:
> In the future maybe we could throw NoSuchMethodError if you try to create a jmethodID out of an obsolete method and remove all this code. But that's not in this change.

Hmmm... I'm rusty with redefinition, but I think there are legitimate scenarios where is "okay"
to call an obsolete method. I believe it has to do with an in-progress call to a method being
redefined and you might get the obsolete method or the new method depending on timing.

dcubed-ojdk commented on src/hotspot/share/oops/instanceKlass.cpp:
```diff
@@ -2470,24 +2471,30 @@ void InstanceKlass::update_methods_jmethod_cache() {
+    if (!m->is_overpass() &&  // skip overpasses
+        id == nullptr) {
+      id = Method::make_jmethod_id(class_loader_data(), m);
+      Atomic::release_store(&jmeths[idnum+1], id);
```
nit spacing: s/size+1/size + 1/

dcubed-ojdk commented on src/hotspot/share/oops/instanceKlass.cpp:
```diff
@@ -2392,12 +2392,16 @@ jmethodID InstanceKlass::update_jmethod_id(jmethodID* jmeths, Method* method, in
+}
+
+
+jmethodID InstanceKlass::get_jmethod_id(Method* method) {
```
Should `method` be `const`?

coleenp replied:
Much of the metadata parameter declarations aren't const-safe.  Adding this one const has a big fall-out leading to some jni calls.  This would be is a good project and we should have it for new code, which some of this is.  I'll see if can add any consts.

dcubed-ojdk replied:
Thanks for investigating!

dcubed-ojdk commented on src/hotspot/share/oops/instanceKlass.hpp:
```diff
@@ -1054,8 +1054,8 @@ class InstanceKlass: public Klass {
   inline void release_set_methods_jmethod_ids(jmethodID* jmeths);
-  // This nulls out jmethodIDs for all methods in 'klass'
-  static void clear_jmethod_ids(InstanceKlass* klass);
+  // This nulls out obsolete jmethodIDs for all methods in 'klass'
```
nit typo: please add an ending period to the comment.

dcubed-ojdk commented on src/hotspot/share/oops/jmethodIDTable.cpp:
```diff
@@ -0,0 +1,185 @@
+#include "logging/log.hpp"
+#include "memory/allocation.hpp"
+#include "memory/resourceArea.hpp"
+#include "oops/method.hpp"
+#include "oops/jmethodIDTable.hpp"
```
Please swap these two #includes into sort order.

dcubed-ojdk commented on src/hotspot/share/oops/jmethodIDTable.cpp:
```diff
@@ -0,0 +1,185 @@
+#include "utilities/concurrentHashTableTasks.inline.hpp"
+#include "utilities/macros.hpp"
+
+// Save (jmethod, Method*) in a hashtable to lookup Method
```
nit typo: please add an ending period to the comment.

dcubed-ojdk commented on src/hotspot/share/oops/jmethodIDTable.cpp:
```diff
@@ -0,0 +1,185 @@
+#include "utilities/macros.hpp"
+
+// Save (jmethod, Method*) in a hashtable to lookup Method
+// The CHT is for performance because it is has lock free lookup.
```
typo: s/because it is has lock free/because it has lock free/

dcubed-ojdk commented on src/hotspot/share/oops/jmethodIDTable.cpp:
```diff
@@ -0,0 +1,185 @@
+  const size_t start_size = 10;
+  // 2^24 is max size
+  const size_t end_size = 24;
+  // If a chain gets to 32 something might be wrong
```
nit typo: please add an ending period to the comment.

dcubed-ojdk commented on src/hotspot/share/oops/jmethodIDTable.cpp:
```diff
@@ -0,0 +1,185 @@
+};
+
+static JmethodEntry* get_jmethod_entry(jmethodID mid) {
+  assert(mid != nullptr, "JNI method id should not be null");
```
Perhaps: s/null/nullptr/
I can't remember if assert failure text output is okay to be `null`.

dcubed-ojdk commented on src/hotspot/share/oops/jmethodIDTable.cpp:
```diff
@@ -0,0 +1,185 @@
+  };
+  bool needs_rehashing = false;
+  _jmethod_id_table->get(current, lookup, get, &needs_rehashing);
+  assert (!needs_rehashing, "should never need rehashing");
```
nit spacing: s/assert (/assert(/

dcubed-ojdk commented on src/hotspot/share/oops/jmethodIDTable.cpp:
```diff
@@ -0,0 +1,185 @@
+
+// Add a method id to the jmethod_ids
+jmethodID JmethodIDTable::make_jmethod_id(Method* m) {
+  bool grow_hint, clean_hint, created;
```
nit: sort local variables?

dcubed-ojdk commented on src/hotspot/share/oops/jmethodIDTable.cpp:
```diff
@@ -0,0 +1,185 @@
+  assert(created, "must be");
+  log_debug(jmethod)("Inserted jmethod id " UINT64_FORMAT_X, _jmethodID_counter);
+
+  // Resize table if it needs to grow.  The _jmethod_id_table has a good distribution
```
nit typo: please add an ending period to the comment.

dcubed-ojdk commented on src/hotspot/share/oops/jmethodIDTable.cpp:
```diff
@@ -0,0 +1,185 @@
+         !_jmethod_id_table->is_max_size_reached());
+}
+
+// Add a method id to the jmethod_ids
```
nit typo: please add an ending period to the comment.

dcubed-ojdk commented on src/hotspot/share/oops/jmethodIDTable.cpp:
```diff
@@ -0,0 +1,185 @@
+    result = value;
+  };
+  bool removed = _jmethod_id_table->remove(current, lookup, get);
+  assert(removed, "should be");
```
"should be" or "must be"?

dcubed-ojdk commented on src/hotspot/share/oops/jmethodIDTable.cpp:
```diff
@@ -0,0 +1,185 @@
+void JmethodIDTable::change_method_associated_with_jmethod_id(jmethodID jmid, Method* new_method) {
+  assert_locked_or_safepoint(JmethodIdCreation_lock);
+  JmethodEntry* result = get_jmethod_entry(jmid);
+  // change to table to point to the new method
```
Perhaps: // Change to table entry to point to the new method.

dcubed-ojdk commented on src/hotspot/share/oops/jmethodIDTable.cpp:
```diff
@@ -0,0 +1,185 @@
+  JmethodEntry* result = get_jmethod_entry(jmid);
+  // We need to make sure that jmethodID actually resolves to this method
+  // - multiple redefined versions may share jmethodID slots and if a method
+  //   has already been rewired to a newer version we could be removing reference
```
typo?: s/could be removing reference/could be clearing a reference/

dcubed-ojdk commented on src/hotspot/share/oops/jmethodIDTable.hpp:
```diff
@@ -0,0 +1,53 @@
+#include "jni.h"
+#include "memory/allocation.hpp"
+
+// Class for associating Method with jmethodID
```
nit typo: please add an ending period to the comment.

dcubed-ojdk commented on src/hotspot/share/oops/jmethodIDTable.hpp:
```diff
@@ -0,0 +1,53 @@
+ public:
+  static void initialize();
+
+  // Given a Method return a jmethodID
```
nit typo: please add an ending period to the comment.

dcubed-ojdk commented on src/hotspot/share/oops/jmethodIDTable.hpp:
```diff
@@ -0,0 +1,53 @@
+  // Given a Method return a jmethodID
+  static jmethodID make_jmethod_id(Method* m);
+
+  // Given a jmethodID, return a Method
```
nit typo: please add an ending period to the comment.

dcubed-ojdk commented on src/hotspot/share/oops/jmethodIDTable.cpp:
```diff
@@ -0,0 +1,185 @@
+  Thread* current = Thread::current();
+  JmethodEntry* result;
+  auto get = [&] (JmethodEntry* value) {
+    // function called if value is found so is never null
```
Perhaps: s/null/nullptr/

dcubed-ojdk commented on src/hotspot/share/oops/jmethodIDTable.hpp:
```diff
@@ -0,0 +1,53 @@
+  static Method* resolve_jmethod_id(jmethodID mid);
+
+  // Class unloading support, remove the associations from the tables.  Stale jmethodID will
+  // not be found and return null
```
nit typo: please add an ending period to the comment.

Perhaps: s/null/nullptr/

dcubed-ojdk commented on src/hotspot/share/oops/jmethodIDTable.cpp:
```diff
@@ -0,0 +1,185 @@
+  JmethodIDLookup lookup((uint64_t)mid);
+  JmethodEntry* result = nullptr;
+  auto get = [&] (JmethodEntry* value) {
+    // function called if value is found so is never null
```
Perhaps: s/null/nullptr/

dcubed-ojdk commented on src/hotspot/share/oops/method.cpp:
```diff
@@ -2059,198 +2060,68 @@ void BreakpointInfo::clear(Method* method) {
-}
+// jmethodIDs are 64-bit integers that will never run out and are mapped in a table
+// to their Method and vice versa.  If JNI code has access to stale jmethodID, this
+// wastes no memory but the Method* returned is null
```
Perhaps: s/null/nullptr/

dcubed-ojdk commented on src/hotspot/share/oops/method.cpp:
```diff
@@ -2059,198 +2060,68 @@ void BreakpointInfo::clear(Method* method) {
-    cld->jmethod_ids()->ensure_methods(capacity);
-  }
-}
+// jmethodIDs are 64-bit integers that will never run out and are mapped in a table
```
Should we have a `guarantee` or `assert` somewhere that the counter never wraps?

dcubed-ojdk replied:
Thanks!

dholmes-ora replied:
Given it will take 584 years to wrap-around at a generation rate of 1 per nanosecond, I think we can just use an assertion here, as the only way this could fire is if we initialize the counter incorrectly.

dcubed-ojdk commented on src/hotspot/share/oops/method.cpp:
```diff
@@ -2059,198 +2060,68 @@ void BreakpointInfo::clear(Method* method) {
-  }
-  // jmethodID is a pointer to Method*
-  return (jmethodID)cld->jmethod_ids()->add_method(m);
+  // Add to growable array in CLD
```
nit typo: please add an ending period to the comment.

dcubed-ojdk commented on src/hotspot/share/oops/method.cpp:
```diff
@@ -2059,198 +2060,68 @@ void BreakpointInfo::clear(Method* method) {
+
+// Get the Method out of the table given the method id.
+Method* Method::resolve_jmethod_id(jmethodID mid) {
+  assert(mid != nullptr, "JNI method id should not be null");
```
Perhaps: s/null/nullptr/

dcubed-ojdk commented on src/hotspot/share/oops/method.hpp:
```diff
@@ -704,29 +704,20 @@ class Method : public Metadata {
-
   // Use resolve_jmethod_id() in situations where the caller is expected
   // to provide a valid jmethodID; the only sanity checks are in asserts;
   // result guaranteed not to be null.
```
Perhaps: s/null/nullptr/

dcubed-ojdk commented on src/hotspot/share/oops/method.hpp:
```diff
@@ -704,29 +704,20 @@ class Method : public Metadata {
+  static Method* resolve_jmethod_id(jmethodID mid);
 
   // Use checked_resolve_jmethod_id() in situations where the caller
   // should provide a valid jmethodID, but might not. Null is returned
```
Perhaps: s/Null/Nullptr/

dcubed-ojdk commented on src/hotspot/share/prims/jvmtiEnv.cpp:
```diff
@@ -2768,9 +2768,11 @@ JvmtiEnv::GetClassMethods(oop k_mirror, jint* method_count_ptr, jmethodID** meth
-  bool jmethodids_found = true;
   int skipped = 0;  // skip overpass methods
 
+  // Make jmethodIDs for all non-overpass methods
```
nit typo: please add an ending period to the comment.

dcubed-ojdk commented on src/hotspot/share/prims/jvmtiEnv.cpp:
```diff
@@ -2783,20 +2785,8 @@ JvmtiEnv::GetClassMethods(oop k_mirror, jint* method_count_ptr, jmethodID** meth
-      id = m->jmethod_id();
-    }
+    jmethodID id = m->find_jmethod_id_or_null();
+    assert (id != nullptr, "should be created above");
```
nit spacing: s/assert (/assert(/

dcubed-ojdk commented on src/hotspot/share/runtime/mutexLocker.cpp:
```diff
@@ -233,7 +233,7 @@ void mutex_init() {
   MUTEX_DEFN(Notification_lock               , PaddedMonitor, service);          // used for notification thread operations
 
-  MUTEX_DEFN(JmethodIdCreation_lock          , PaddedMutex  , nosafepoint-2); // used for creating jmethodIDs.
+  MUTEX_DEFN(JmethodIdCreation_lock          , PaddedMutex  , nosafepoint-1); // used for creating jmethodIDs.
```
Interesting. Why change from `nosafepoint-2` to `nosafepoint-1`?

coleenp replied:
Found why it needed to be nosafepoint-1.  The DEFL macro is good for the global locks but HandshakeState_lock is not global since there's one per handshake operation.

coleenp commented:
There is a way to create a jmethodID from an obsolete method and I thought we had tests that did this. I'm not finding them right now. The old and obsolete methods can still be running after a redefinition if they were running during the redefinition.  The JIT deoptimizes them but they will be run in the interpreter.  New invocations of the method will choose the new version of the method always.  Always unless there's a bug that we don't know about.  We have fixed a few old method invocations in the past coming from various places in the JVM but we fixed the "last" one fairly recently.
But technically, one could create a jmethodID to an obsolete method although it's not easy to do.  I was musing above but not committed to breaking compatibility if there is code that requires this.

dcubed-ojdk commented on src/hotspot/share/classfile/classLoaderData.cpp:
```diff
@@ -578,6 +579,32 @@ void ClassLoaderData::remove_class(Klass* scratch_class) {
+// releases memory.
+// Because native code (e.g. JVMTI agent) holding jmethod_ids may access them
+// after the associated classes and class loader are unloaded, subsequent lookups
+// for these ids will return null since they are no longer found in the table.
```
I'm not sure, but your reasoning sounds good to me!

dcubed-ojdk reviewed (approved):
Thumbs up with the latest version!

Thanks for fixing all the nits.

dholmes-ora reviewed (commented):
This looks good. I'm still digesting it.

Thanks

xmas92 commented on src/hotspot/share/oops/jmethodIDTable.cpp:
```diff
@@ -0,0 +1,186 @@
+}
+
+static bool needs_resize(Thread* current) {
+  return ((_jmethodID_counter > (_resize_load_trigger * table_size(current))) &&
+         !_jmethod_id_table->is_max_size_reached());
```
Should we not just have a separate jmethodID entry count variable we use here instead, that is incremented and decremented on insert and remove. Rather than using the next jmethodID counter which just grows monotonically regardless of and removals.

sspitsyn replied:
Interesting suggestion to consider. I'm not sure yet if it is really important.

coleenp replied:
This makes sense because we want to know how many items are in the table vs. how many items have been in the table.

dholmes-ora commented on src/hotspot/share/oops/instanceKlass.cpp:
```diff
@@ -2470,24 +2471,30 @@ void InstanceKlass::update_methods_jmethod_cache() {
-  int new_jmeths = 0;
+void InstanceKlass::make_methods_jmethod_ids() {
+  MutexLocker ml(JmethodIdCreation_lock, Mutex::_no_safepoint_check_flag);
+  jmethodID* jmeths = methods_jmethod_ids_acquire();
```
Acquire is only ever needed outside the lock. I don't like there only being acquire/release available generally speaking because it just confuses what memory operations are being synchronized.

sspitsyn commented on src/hotspot/share/oops/jmethodIDTable.hpp:
```diff
@@ -0,0 +1,53 @@
+  // not be found and return null.
+  static void remove(jmethodID mid);
+
+  // RedefineClasses support
```
Nit: Add a dot at the end of the comment for consistency with other comments.

coleenp replied:
This isn't a sentence and this pattern is in the sources in a lot of places, none of these places have a period at the end.

sspitsyn commented on src/hotspot/share/oops/instanceKlass.cpp:
```diff
@@ -2391,13 +2391,26 @@ jmethodID InstanceKlass::update_jmethod_id(jmethodID* jmeths, Method* method, in
+  Atomic::release_store(&_methods_jmethod_ids, jmeths);
+}
+
+// Lookup or create a jmethodID
```
Nit: Add dot at the end.

sspitsyn reviewed (commented):
This is really great optimization and refactoring!
Looks pretty good to me but I'd like to make one more pass through the changes.

coleenp commented on src/hotspot/share/oops/method.hpp:
```diff
@@ -704,29 +704,20 @@ class Method : public Metadata {
 
   static void change_method_associated_with_jmethod_id(jmethodID old_jmid_ptr, Method* new_method);
-  static bool is_method_id(jmethodID mid);
+  static bool validate_method_id(jmethodID mid);
```
Good suggestion - these names get confusing without the 'j'.  I tried to make them consistent.

dholmes-ora reviewed (commented):
Still working my way through this ...

Just to be sure I have the correct mental picture of this, things work as follow:

- a `jMethodID` is just a unique integer index to represent a `Method` for use with JNI
- when we allocate a new `jMethodID` to a `Method` we 
  - add that mapping to the global table
  - add an entry to the `ClassLoaderData`'s `jMethodID` "list"
- each `instanceklass` also maintains a "cache" of its own `jMethodID` mappings (some of which need updating on method redefinition)
- when a class is unloaded we deallocate the cache, deallocate the CLD list, and remove the table entries
- when a JNI API is presented with a `jMethodID` by the caller, it validates it by looking it up in the table, to see if the mapping exists

The cache is created under a lock, and `jMethodID`s are created and added to the cache (and the table, and the CLD list) under the same lock. But access to the cache is lock-free using acquire/release.

The cache is also destroyed under the same lock, and the table entries removed under that same lock, and the CLD list deallocated.

Is the above correct? What details have I missed?

Thanks

coleenp commented:
@tstuefe I think your method would still have a race and still allow a stale pointer crash, but just less likely, with more code to manage how to delete these pointers.

The memory leak for jmethodIDs has been observed in our testing, resulting in native OOMs.  We discussed a few things to solve this in the bugs, and this was the best idea we had at the time.  It is a big change though.

dholmes-ora commented:
> I think this should prevent this race but I'd have to think about it some more.

Aside from the fact the `checked` method is not used much, the problem is that if the caller does not have something keeping the class alive, then the resolution of the jMethodID can succeed and we will proceed with trying to call the method. In the meantime the fact the class is unreferenced could be noticed and the class then unload. Now that can only happen at safepoints, so it then depends on the details of the code that tries to invoke the method e.g. in jni.cpp
```
static void jni_invoke_static(JNIEnv *env, JavaValue* result, jobject receiver, JNICallType call_type, jmethodID method_id, JNI_ArgumentPusher *args, TRAPS) {
  methodHandle method(THREAD, Method::resolve_jmethod_id(method_id));

  // Create object to hold arguments for the JavaCall, and associate it with
  // the jni parser
  ResourceMark rm(THREAD);
  int number_of_parameters = method->size_of_parameters();
  JavaCallArguments java_args(number_of_parameters);

  assert(method->is_static(), "method should be static");

  // Fill out JavaCallArguments object
  args->push_arguments_on(&java_args);
  // Initialize result type
  result->set_type(args->return_type());

  // Invoke the method. Result is returned as oop.
  JavaCalls::call(result, method, &java_args, CHECK);
```
Can we hit safepoint checks anywhere on the path to the actual invocation of the method? If not, what is guaranteeing that?

coleenp commented:
I'm not sure but I think in this instance, the native caller has the mirror for the class where it gets the jmethodID from, so can't unload the Method.

```
JNI_ENTRY(ResultType, \
          jni_CallStatic##Result##Method(JNIEnv *env, jclass cls, jmethodID methodID, ...)) \
\
```

dcubed-ojdk reviewed (approved):
Thumbs up. Reviewed the changes between v02 and v11
and the changed look good.

dcubed-ojdk commented:
> > I think this should prevent this race but I'd have to think about it some more.
> 
> Aside from the fact the `checked` method is not used much, the problem is
> that if the caller does not have something keeping the class alive, then the
> esolution of the jMethodID can succeed and we will proceed with trying to
> call the method. In the meantime the fact the class is unreferenced could be
> noticed and the class then unloaded. Now that can only happen at safepoints,
> so it then depends on the details of the code that tries to invoke the method
>  e.g. in jni.cpp

Consider this code from above:
> static void jni_invoke_static(JNIEnv *env, JavaValue* result, jobject receiver, JNICallType call_type, jmethodID method_id, JNI_ArgumentPusher *args, TRAPS) {
>   methodHandle method(THREAD, Method::resolve_jmethod_id(method_id));

Once `Method::resolve_jmethod_id(method_id)` returns to the caller, how can
the class be unreferenced?

dholmes-ora commented:
> I'm not sure but I think in this instance, the native caller has the mirror for the class where it gets the jmethodID from, so can't unload the Method.
> 
> ```
> JNI_ENTRY(ResultType, \
>           jni_CallStatic##Result##Method(JNIEnv *env, jclass cls, jmethodID methodID, ...)) \
> \
> ```

@coleenp Take a look a the complete code:
```
JNI_ENTRY(ResultType, \
          jni_CallStatic##Result##Method(JNIEnv *env, jclass cls, jmethodID methodID, ...)) \
\
  EntryProbe; \
  ResultType ret{}; \
  DT_RETURN_MARK_FOR(Result, CallStatic##Result##Method, ResultType, \
                     (const ResultType&)ret);\
\
  va_list args; \
  va_start(args, methodID); \
  JavaValue jvalue(Tag); \
  JNI_ArgumentPusherVaArg ap(methodID, args); \
  jni_invoke_static(env, &jvalue, nullptr, JNI_STATIC, methodID, &ap, CHECK_(ResultType{})); \
  va_end(args); \
  ret = jvalue.get_##ResultType(); \
  return ret;\
JNI_END
```
the `cls` parameter is never actually used. So while it is supposed to refer to the class you have the static method jMethodID for, there is no requirement that it actually does, and could even be null.

> Once Method::resolve_jmethod_id(method_id) returns to the caller, how can the class be unreferenced?

@dcubed-ojdk  because the caller does not hold a reference to it initially, and resolving the jMethodID does not create a new reference to it. So once we have resolved, at the next safepoint the class could be seen as "not alive" and at the safepoint after that it can be unloaded. Hence we are relying on there being no safepoint checks in the upcall code, after the resolution,  to ensure this can't happen.

xmas92 reviewed (approved):
Looks good. Thanks.

_Feels like `JmethodIdCreation_lock` might not be needed now but it is also easier to reason about things when we serialize the modifications._

coleenp commented:
Thank you for reviewing from the GC and class unloading perspective, Axel and Erik.

coleenp commented:
Thank you Serguei.

dcubed-ojdk reviewed (approved):
I've made another pass through this PR. Just a few more nits.

dcubed-ojdk commented:
Ping @jbachorik for discussions about the safety of jmethodIDs...

coleenp commented on src/hotspot/share/runtime/mutexLocker.cpp:
```diff
@@ -233,7 +233,7 @@ void mutex_init() {
   MUTEX_DEFN(Notification_lock               , PaddedMonitor, service);          // used for notification thread operations
 
-  MUTEX_DEFN(JmethodIdCreation_lock          , PaddedMutex  , nosafepoint-2); // used for creating jmethodIDs.
+  MUTEX_DEFN(JmethodIdCreation_lock          , PaddedMutex  , nosafepoint-1);    // used for creating jmethodIDs locks HandshakeState_lock
```
ugh that makes the line too long for me.

dcubed-ojdk replied:
How about just: s/locks/can also lock/

coleenp replied:
Done - that's not too wide for my screen.

dcubed-ojdk reviewed (approved):
Thanks for making the wording tweaks.

dholmes-ora reviewed (approved):
Still fine
